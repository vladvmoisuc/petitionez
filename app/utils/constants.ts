export const DATE_FORMATS = {
  STANDARD: "D/MM/YYYY",
};

export const ROUTES = {
  MANAGEMENT: "/management",
  CAMPAIGNS: "/campaigns",
  LOGIN: "/login",
};

export const EMAIL_TYPES = {
  NEW: "new",
  OLD: "old",
};

export const LOGIN_STATES = {
  ...EMAIL_TYPES,
  DEFAULT: "default",
  ERROR: "error",
};

export const INTENTS = {
  VALIDATE: "validate",
  GENERATE: "generate",
  DELETE: {
    CAMPAIGN: "delete-campaign",
  },
  UPDATE: {
    CAMPAIGN: "update-campaign",
  },
  CREATE: {
    ENTITY: "create-entity",
    CAMPAIGN: "create-campaign",
  },
};

export const SLUGS = {
  CREATE: "create",
};

export const LINKS = {
  POSTIMAGES: {
    URL: "https://postimages.org/",
    NAME: "postimages.org",
  },
};
