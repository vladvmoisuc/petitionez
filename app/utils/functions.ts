import ro from "~/content/ro.json";
import errors from "~/content/errors.json";
import logs from "~/content/logs.json";

import type { DynamicObject } from "~/utils/types";

export function waitFor(time: number = 500) {
  return new Promise<void>((resolve) =>
    setTimeout(() => {
      resolve();
    }, time)
  );
}

export function replace(data: { [key: string]: string }, content: string) {
  return Object.keys(data).reduce(
    (text, key) => text.replace(new RegExp(`{{${key}}}`, "g"), data[key]),
    content
  );
}

function extractContent(
  source: DynamicObject,
  key: string,
  values?: DynamicObject
): string {
  const content = source[key as keyof typeof source] ?? "";

  return values ? replace(values, content) : content;
}

export function getText(key: string, values?: DynamicObject): string {
  return extractContent(ro, key, values);
}

export function getError(key: string, values?: DynamicObject): string {
  return extractContent(errors, key, values);
}

export function getLog(key: string, values?: DynamicObject): string {
  return extractContent(logs, key, values);
}

export async function getFormData(
  request: Request
): Promise<{ keys: string[]; values: DynamicObject }> {
  const data = await request.formData();
  const keys = [...data.keys()];

  const values = keys.reduce((values, key) => {
    return {
      ...values,
      [key]: data.get(key),
    };
  }, {});

  return {
    keys,
    values,
  };
}
