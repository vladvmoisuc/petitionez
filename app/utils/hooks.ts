import { useState, useEffect, useCallback } from "react";

import type { Transition } from "@remix-run/react/dist/transition";

export const useLocalStorage = <T>(
  key: string
): [T | undefined, typeof setLocalStorage] => {
  const [state, setState] = useState();

  const isClientSide = typeof window !== "undefined";

  useEffect(() => {
    const value = window.localStorage.getItem(key);

    if (isClientSide && value) {
      setState(typeof value === "string" ? JSON.parse(value) : value);
    }
  }, [isClientSide, key]);

  const setLocalStorage = useCallback(
    (data: any) => {
      const value = typeof data !== "string" ? JSON.stringify(data) : data;

      if (isClientSide) {
        window.localStorage.setItem(key, value);

        setState({ ...data });
      }
    },
    [isClientSide, key]
  );

  return [state, setLocalStorage];
};

export const useAlert = (transition: Transition, time: number = 1000) => {
  const [alert, setAlert] = useState(false);

  useEffect(() => {
    if (transition.state === "loading" && transition.type === "actionReload") {
      setAlert(true);

      setTimeout(() => {
        setAlert(false);
      }, time);
    }
  }, [transition.state, transition.type, time]);

  return alert;
};
