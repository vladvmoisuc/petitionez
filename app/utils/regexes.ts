export default {
  EMAIL: /^\S+@\S+\.\S+$/,
  URL: /^(http:\/\/|https:\/\/).+\..+/,
  START_LETTER: /^[a-z]/g,
  DASHES: /-/g,
  SPACES: /\s/g,
};
