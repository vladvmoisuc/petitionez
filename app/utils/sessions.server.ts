import { createCookieSessionStorage } from "@remix-run/node";

const HOUR = 1 * 60 * 60; // second -> minute -> hour

const { getSession, commitSession, destroySession } =
  createCookieSessionStorage({
    cookie: {
      name: "__session",
      httpOnly: true,
      maxAge: HOUR,
      path: "/",
      sameSite: "lax",
      secrets: [process.env.COOKIES_SECRET_KEY as string],
      secure: true,
    },
  });

async function getLoggedUser(request: Request) {
  const session = await getSession(request.headers.get("Cookie"));

  const loggedUserEmail: string = session.get("email") ?? "";
  const loggedUserId: string = session.get("id") ?? "";
  const isAdmin: boolean = session.get("admin") ?? false;

  return { loggedUserEmail, loggedUserId, isAdmin };
}

export { getSession, commitSession, destroySession, getLoggedUser };
