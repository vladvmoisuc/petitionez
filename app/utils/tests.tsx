import { createMemoryRouter, RouterProvider } from "react-router-dom";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import type { ReactElement, ReactNode } from "react";

export function renderWithUser(jsx: ReactElement) {
  return {
    user: userEvent.setup({ delay: null }),
    ...render(jsx),
  };
}

export function renderWithRouter(jsx: ReactElement) {
  const Wrapper = ({ children }: { children: ReactNode }) => (
    <RouterProvider
      router={createMemoryRouter([
        {
          path: "/",
          element: children,
        },
      ])}
    />
  );

  return {
    user: userEvent.setup({ delay: null }),
    ...render(jsx, {
      wrapper: Wrapper,
    }),
  };
}
