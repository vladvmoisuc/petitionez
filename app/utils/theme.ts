import { createTheme, colors } from "@mui/material";

const FONTS = {
  PLAYFAIR_DISPLAY: "'Playfair Display', serif",
  RED_HAT: "'Red Hat Mono', monospace",
};

export default createTheme({
  typography: {
    h1: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h2: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h3: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h4: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    h5: {
      fontFamily: FONTS.PLAYFAIR_DISPLAY,
    },
    fontFamily: FONTS.RED_HAT,
  },
  components: {
    MuiAppBar: {
      defaultProps: {
        color: "transparent",
        elevation: 0,
        position: "static",
      },
    },
    MuiToolbar: {
      styleOverrides: {
        root: {
          justifyContent: "space-between",
          borderBottomWidth: 1,
          borderBottomStyle: "solid",
          borderBottomColor: colors.grey[500],
        },
      },
    },
    MuiCard: {
      defaultProps: {
        elevation: 0,
        square: true,
      },
    },
    MuiSkeleton: {
      styleOverrides: {
        root: {
          backgroundColor: colors.grey[100],
        },
      },
    },
    MuiContainer: {
      defaultProps: {
        maxWidth: false,
      },
    },
  },
});
