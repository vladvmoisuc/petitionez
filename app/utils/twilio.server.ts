import sendgrid from "@sendgrid/mail";
import { readFileSync } from "fs";
import { join } from "path";

import { replace } from "~/utils/functions";

sendgrid.setApiKey(process.env.SENDGRID_API_KEY as string);

const TEMPLATES = {
  OTP: {
    from: "petitionez@gmail.com",
    subject: "Cod de autentificare petitionez.ro",
    html: readFileSync(
      join(__dirname, "..", "app", "templates", "emails", "otp.html")
    ).toString(),
  },
};

type Params = {
  type: keyof typeof TEMPLATES;
  to: string;
  data: {
    [key: string]: string;
  };
};

export default {
  send: async ({ type, to, data = {} }: Params) =>
    await sendgrid.send({
      ...TEMPLATES[type],
      to,
      html: replace(data, TEMPLATES[type].html),
    }),
};
