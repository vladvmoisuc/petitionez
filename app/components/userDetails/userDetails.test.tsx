import { render, screen } from "@testing-library/react";

import UserDetails from ".";

import { getText } from "~/utils/functions";
import { renderWithUser } from "~/utils/tests";

const props = {
  isMobile: false,
  error: false,
  user: {
    name: "John",
    address: "Sesame Street",
  },
  onChange: jest.fn(),
};

describe("<UserDetails />", () => {
  it("should match the snapshot", () => {
    const { container } = render(<UserDetails {...props} />);

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot with mobile content", () => {
    const { container } = render(<UserDetails {...props} isMobile />);

    expect(container).toMatchSnapshot();
  });

  it("should match the snapshot with errors", () => {
    const { container } = render(
      <UserDetails {...props} error user={{ name: "", address: "" }} />
    );

    expect(container).toMatchSnapshot();
  });

  it("should call the onChange prop when filling the fields", async () => {
    const { user } = renderWithUser(
      <UserDetails {...props} user={{ name: "", address: "" }} />
    );

    await user.type(
      screen.getByLabelText(getText("campain.user.form.fields.name")),
      "F"
    );

    await user.type(
      screen.getByLabelText(getText("campain.user.form.fields.address")),
      "N"
    );

    expect(props.onChange).toBeCalledTimes(2);
  });
});
