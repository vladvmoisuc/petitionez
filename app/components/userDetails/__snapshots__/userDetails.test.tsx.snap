// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDetails /> should match the snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1hu82mc-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-8kjoqe"
    >
      <div
        class="css-m69qwo-MuiStack-root"
      >
        <span
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          Date personale
          .
        </span>
        <svg
          aria-hidden="true"
          aria-label="Datele tale nu sunt salvate în baza noastră de date. Singurul loc în care sunt păstrate e memoria locală a browserului tău, ce poate fi ștearsă de către tine în orice moment."
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="InfoOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r1:"
          id=":r1:-label"
        >
          Nume și prenume
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id=":r1:"
            name="name"
            type="text"
            value="John"
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r2:"
          id=":r2:-label"
        >
          Adresa(întreagă) din buletin
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id=":r2:"
            name="address"
            type="text"
            value="Sesame Street"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<UserDetails /> should match the snapshot with errors 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1hu82mc-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-8kjoqe"
    >
      <div
        class="css-m69qwo-MuiStack-root"
      >
        <span
          class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
        >
          Date personale
          .
        </span>
        <svg
          aria-hidden="true"
          aria-label="Datele tale nu sunt salvate în baza noastră de date. Singurul loc în care sunt păstrate e memoria locală a browserului tău, ce poate fi ștearsă de către tine în orice moment."
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="InfoOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r6:"
          id=":r6:-label"
        >
          Nume și prenume
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="true"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id=":r6:"
            name="name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r7:"
          id=":r7:-label"
        >
          Adresa(întreagă) din buletin
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
        >
          <input
            aria-invalid="true"
            class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
            id=":r7:"
            name="address"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<UserDetails /> should match the snapshot with mobile content 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r3:"
      id=":r3:-label"
    >
      Nume și prenume
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        id=":r3:"
        name="name"
        type="text"
        value="John"
      />
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r4:"
      id=":r4:-label"
    >
      Adresa(întreagă) din buletin
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-l4u8b9-MuiInputBase-root-MuiInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
        id=":r4:"
        name="address"
        type="text"
        value="Sesame Street"
      />
    </div>
  </div>
</div>
`;
